### ISSUE TITLE
Buffer appears to be growing when required_acks set to -1

### ISSUE THREAD

**Issue by dtboctor** [fluent/fluent-plugin-kafka#101]
I have a fluentd instance I'm using with the buffered plugin writing to a 3 node cluster with min ISR set to 2, if I run with required_acks set to 1 I have no problem and the buffer stays small, regardless of how large I set the flush interval - I was able to use up to 10 seconds and there's no issue.  If I set the required_acks to -1, The buffer starts to slowly grow, even though looking at the log it appears to be writing frequently, as quickly as once a second.   Is there something else that can be wrong?  I don't see any metrics in Kafka that suggest more than a few millisecond performance hit with the additional setting.

**Comment by repeatedly** [comment]
> Is there something else that can be wrong?

It seems normal behaviour. `-1` means kafka client waits all acks from kafka.
So getting response from kafka takes longer time than 1 or 2.

**Comment by dtboctor** [comment]
Yeah, that makes sense to me, it takes a bit longer to do a produce request, but I guess my confusion is, if I can keep up to date with the buffer when I only produce once every 10 seconds, shouldn't it still be able to stay up to date even if the production request takes a little bit longer if I try producing every second?

**Comment by dtboctor** [comment]
Would changing my buffer_chunk_limit affect the performance of how many records are written per sync to kafka?

**Comment by repeatedly** [comment]
Currently, no. You also need to change `kafka_agg_max_bytes` parameter. https://github.com/fluent/fluent-plugin-kafka/blob/0633788b5d5cd5cee9ac298e84c1f4ae2c2cd761/lib/fluent/plugin/out_kafka_buffered.rb#L250
I will remove this parameter from `out_kafka2`. `out_kafka2` is based on v0.14 API based plugin unlike `out_kafka_buffered`.

**Comment by dtboctor** [comment]
is out_kafka2 a newer version?  How will that one work?  

I noticed that even though I have a chunk size of 128m, running with acks=-1 makes the buffer grow with lots of smaller files (maybe 100k each)  that don't get deleted, can you explain to me the sequence of steps that might be causing that?

**Comment by repeatedly** [comment]
> can you explain to me the sequence of steps that might be causing that?

Could you show me your configuation? I don't have a enough information for that...

**Comment by dtboctor** [comment]
```<match events.production>
  type kafka_buffered
  brokers analytics-kafka01:9092,analytics-kafka02:9092,analytics-kafka03:9092

  default_topic 'events'
  buffer_type file
  buffer_path '/var/log/events/fluentd/events_production/'

  required_acks -1
  flush_interval 1

  buffer_chunk_limit 128m
  buffer_queue_limit 65536
</match>
```

That's my configuration.  When I let this go I start getting a huge collection of file build up each of which is very small: 

```
-rw-r--r--. 1 root root  46K Jan 13 11:06 buffer.b545fe87e45653f9b0a48bf14343e23b1.log
-rw-r--r--. 1 root root   86 Jan 13 11:06 buffer.b545fe87e45653f9b0a48bf14343e23b1.log.meta
-rw-r--r--. 1 root root 106K Jan 13 11:05 buffer.q545fe84f522cb01a438b4966d0eec060.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe84f522cb01a438b4966d0eec060.log.meta
-rw-r--r--. 1 root root 121K Jan 13 11:05 buffer.q545fe8504728ee35361c2d03c1bb45b3.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe8504728ee35361c2d03c1bb45b3.log.meta
-rw-r--r--. 1 root root  90K Jan 13 11:05 buffer.q545fe8513ccdeb67642a09bdb6da3cd3.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe8513ccdeb67642a09bdb6da3cd3.log.meta
-rw-r--r--. 1 root root  81K Jan 13 11:05 buffer.q545fe8524ad55a97c7a2dd1b0c81bbc6.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe8524ad55a97c7a2dd1b0c81bbc6.log.meta
-rw-r--r--. 1 root root 104K Jan 13 11:05 buffer.q545fe853269b47deced924addcf39969.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe853269b47deced924addcf39969.log.meta
-rw-r--r--. 1 root root 104K Jan 13 11:05 buffer.q545fe8541de14af6bfe768cdef5c4e33.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe8541de14af6bfe768cdef5c4e33.log.meta
-rw-r--r--. 1 root root 100K Jan 13 11:05 buffer.q545fe8551260348d0d68fd30cf758bce.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe8551260348d0d68fd30cf758bce.log.meta
-rw-r--r--. 1 root root  88K Jan 13 11:05 buffer.q545fe85607fb587c1de67770cdc0a9d4.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe85607fb587c1de67770cdc0a9d4.log.meta
-rw-r--r--. 1 root root  99K Jan 13 11:05 buffer.q545fe85701056fd06cb6c55498ab2c12.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe85701056fd06cb6c55498ab2c12.log.meta
-rw-r--r--. 1 root root  97K Jan 13 11:05 buffer.q545fe857f448254d3eeebe54addd3ed5.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe857f448254d3eeebe54addd3ed5.log.meta
-rw-r--r--. 1 root root 100K Jan 13 11:05 buffer.q545fe858ec77b1fe762d604efa5e7b74.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe858ec77b1fe762d604efa5e7b74.log.meta
-rw-r--r--. 1 root root 119K Jan 13 11:05 buffer.q545fe859ddc69733798fa04349532f8a.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe859ddc69733798fa04349532f8a.log.meta
-rw-r--r--. 1 root root 112K Jan 13 11:05 buffer.q545fe85ad8fb1d0724876cb3d42c9a1d.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe85ad8fb1d0724876cb3d42c9a1d.log.meta
-rw-r--r--. 1 root root 144K Jan 13 11:05 buffer.q545fe85bc854ad8f59421b8c281e8827.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe85bc854ad8f59421b8c281e8827.log.meta
-rw-r--r--. 1 root root 112K Jan 13 11:05 buffer.q545fe85cbf7fff9582d6e533d891e38f.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe85cbf7fff9582d6e533d891e38f.log.meta
-rw-r--r--. 1 root root 121K Jan 13 11:05 buffer.q545fe85db29e1e1865a0c70e44282eb1.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe85db29e1e1865a0c70e44282eb1.log.meta
-rw-r--r--. 1 root root  90K Jan 13 11:05 buffer.q545fe85ea73e4c95ba9f135aff81b25f.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe85ea73e4c95ba9f135aff81b25f.log.meta
-rw-r--r--. 1 root root  92K Jan 13 11:05 buffer.q545fe85f9dc2111de01df1d0a70b504c.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe85f9dc2111de01df1d0a70b504c.log.meta
-rw-r--r--. 1 root root  66K Jan 13 11:05 buffer.q545fe86093b642894384e8d7a8a689fb.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe86093b642894384e8d7a8a689fb.log.meta
-rw-r--r--. 1 root root 107K Jan 13 11:05 buffer.q545fe8618a82b4d365b743bd47179a79.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe8618a82b4d365b743bd47179a79.log.meta
-rw-r--r--. 1 root root 125K Jan 13 11:05 buffer.q545fe8627e3666f16721bac7be346923.log
-rw-r--r--. 1 root root   86 Jan 13 11:05 buffer.q545fe8627e3666f16721bac7be346923.log.meta
-rw-r--r--. 1 root root  71K Jan 13 11:06 buffer.q545fe8636f6d54c2f1dd2e38dea48c09.log
-rw-r--r--. 1 root root   86 Jan 13 11:06 buffer.q545fe8636f6d54c2f1dd2e38dea48c09.log.meta
-rw-r--r--. 1 root root 127K Jan 13 11:06 buffer.q545fe86463f08d5fb9f46403e289ee90.log
-rw-r--r--. 1 root root   86 Jan 13 11:06 buffer.q545fe86463f08d5fb9f46403e289ee90.log.meta
-rw-r--r--. 1 root root 112K Jan 13 11:06 buffer.q545fe865654580ca40589ff790221144.log
-rw-r--r--. 1 root root   86 Jan 13 11:06 buffer.q545fe865654580ca40589ff790221144.log.meta
-rw-r--r--. 1 root root 144K Jan 13 11:06 buffer.q545fe8664e2dc865990d9cf32bd4c946.log
-rw-r--r--. 1 root root   87 Jan 13 11:06 buffer.q545fe8664e2dc865990d9cf32bd4c946.log.meta
-rw-r--r--. 1 root root 116K Jan 13 11:06 buffer.q545fe86744c7e316f22e1e14400f6145.log
-rw-r--r--. 1 root root   86 Jan 13 11:06 buffer.q545fe86744c7e316f22e1e14400f6145.log.meta
-rw-r--r--. 1 root root 148K Jan 13 11:06 buffer.q545fe868433a48b7fb707c6944a1e27f.log
-rw-r--r--. 1 root root   87 Jan 13 11:06 buffer.q545fe868433a48b7fb707c6944a1e27f.log.meta
```

If I change the `required_acks = 1` it works fine, regardless of the `flush_interval`.  I guess what I am saying is that I know that the required acks probably takes an extra 50-100 milliseconds to commit, but it still should provide a flush frequency that should allow this to work, and based on all the tiny queued files I think something else is going on that is not correct.

**Comment by repeatedly** [comment]
Hmm... It seems `required_ack -1` takes 1+ seconds to get all acks from kafka.
So buffer is growing. I'm not sure how to optimize ack speed but tuning kafka side is needed.

**Comment by dtboctor** [comment]
So, I can reproduce the same issue with flush_interval  = 10, in which case the 1 second acks shouldn't be an issue? 

```
2017-01-13_20:29:57.87826 2017-01-13 12:29:57 -0800 [debug]: plugin/buffer.rb:403:block in purge_chunk: chunk purged instance=70086589215680 chunk_id="545ffaed5dcb7bdaaef67d8d3edb318b" metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}>
2017-01-13_20:29:57.88057 2017-01-13 12:29:57 -0800 [debug]: plugin/buffer.rb:358:dequeue_chunk: dequeueing a chunk instance=70086589215680
2017-01-13_20:29:57.88068 2017-01-13 12:29:57 -0800 [debug]: plugin/buffer.rb:367:block in dequeue_chunk: chunk dequeued instance=70086589215680 metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}>
2017-01-13_20:29:57.88076 2017-01-13 12:29:57 -0800 [debug]: plugin/output.rb:787:try_flush: trying flush for a chunk chunk="545ffaee52934e16df12882678582c79"
2017-01-13_20:29:58.11731 2017-01-13 12:29:58 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086589414060
2017-01-13_20:29:58.11733 2017-01-13 12:29:58 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086587412660
2017-01-13_20:29:58.11733 2017-01-13 12:29:58 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086589215680
2017-01-13_20:29:58.11734 2017-01-13 12:29:58 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086588950900
2017-01-13_20:29:58.11734 2017-01-13 12:29:58 -0800 [debug]: plugin/buffer.rb:305:enqueue_chunk: enqueueing chunk instance=70086589215680 metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}>
2017-01-13_20:29:58.38858 2017-01-13 12:29:58 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086587269840
2017-01-13_20:29:59.11912 2017-01-13 12:29:59 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086589414060
2017-01-13_20:29:59.11917 2017-01-13 12:29:59 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086587412660
2017-01-13_20:29:59.11929 2017-01-13 12:29:59 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086589215680
2017-01-13_20:29:59.11938 2017-01-13 12:29:59 -0800 [debug]: plugin/buffer.rb:305:enqueue_chunk: enqueueing chunk instance=70086589215680 metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}>
2017-01-13_20:29:59.11976 2017-01-13 12:29:59 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086588950900
2017-01-13_20:30:00.12851 2017-01-13 12:30:00 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086589414060
2017-01-13_20:30:00.12853 2017-01-13 12:30:00 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086587412660
2017-01-13_20:30:00.12853 2017-01-13 12:30:00 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086588950900
2017-01-13_20:30:00.12854 2017-01-13 12:30:00 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086589215680
2017-01-13_20:30:00.12854 2017-01-13 12:30:00 -0800 [debug]: plugin/buffer.rb:305:enqueue_chunk: enqueueing chunk instance=70086589215680 metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}>
2017-01-13_20:30:00.23635 2017-01-13 12:30:00 -0800 [debug]: plugin/out_kafka_buffered.rb:236:write: (records|bytes) ({"events"=>98}|{"events"=>135606})
2017-01-13_20:30:00.23646 2017-01-13 12:30:00 -0800 [debug]: plugin/buffer.rb:390:block in purge_chunk: purging a chunk instance=70086589215680 chunk_id="545ffaee52934e16df12882678582c79" metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}>
2017-01-13_20:30:00.23718 2017-01-13 12:30:00 -0800 [debug]: plugin/buffer.rb:403:block in purge_chunk: chunk purged instance=70086589215680 chunk_id="545ffaee52934e16df12882678582c79" metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}>
2017-01-13_20:30:00.23981 2017-01-13 12:30:00 -0800 [debug]: plugin/buffer.rb:358:dequeue_chunk: dequeueing a chunk instance=70086589215680
2017-01-13_20:30:00.23990 2017-01-13 12:30:00 -0800 [debug]: plugin/buffer.rb:367:block in dequeue_chunk: chunk dequeued instance=70086589215680 metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}>
2017-01-13_20:30:00.23995 2017-01-13 12:30:00 -0800 [debug]: plugin/output.rb:787:try_flush: trying flush for a chunk chunk="545ffaef470a491b040a85e6801a1733"
2017-01-13_20:30:01.13748 2017-01-13 12:30:01 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086587412660
2017-01-13_20:30:01.13750 2017-01-13 12:30:01 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086589414060
2017-01-13_20:30:01.13750 2017-01-13 12:30:01 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086589215680
2017-01-13_20:30:01.13750 2017-01-13 12:30:01 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086588950900
2017-01-13_20:30:01.13764 2017-01-13 12:30:01 -0800 [debug]: plugin/buffer.rb:305:enqueue_chunk: enqueueing chunk instance=70086589215680 metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}>
2017-01-13_20:30:01.96335 2017-01-13 12:30:01 -0800 [debug]: plugin/out_kafka_buffered.rb:236:write: (records|bytes) ({"events"=>80}|{"events"=>119889})
2017-01-13_20:30:01.96354 2017-01-13 12:30:01 -0800 [debug]: plugin/buffer.rb:390:block in purge_chunk: purging a chunk instance=70086589215680 chunk_id="545ffaef470a491b040a85e6801a1733" metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}>
2017-01-13_20:30:01.96592 2017-01-13 12:30:01 -0800 [debug]: plugin/buffer.rb:403:block in purge_chunk: chunk purged instance=70086589215680 chunk_id="545ffaef470a491b040a85e6801a1733" metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}>
2017-01-13_20:30:01.97068 2017-01-13 12:30:01 -0800 [debug]: plugin/buffer.rb:358:dequeue_chunk: dequeueing a chunk instance=70086589215680
2017-01-13_20:30:01.97082 2017-01-13 12:30:01 -0800 [debug]: plugin/buffer.rb:367:block in dequeue_chunk: chunk dequeued instance=70086589215680 metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}>
2017-01-13_20:30:01.97092 2017-01-13 12:30:01 -0800 [debug]: plugin/output.rb:787:try_flush: trying flush for a chunk chunk="545ffaf0490f94f5ab7391b43eb50bea"
2017-01-13_20:30:02.14523 2017-01-13 12:30:02 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086588950900
2017-01-13_20:30:02.14526 2017-01-13 12:30:02 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086589414060
2017-01-13_20:30:02.14526 2017-01-13 12:30:02 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086589215680
2017-01-13_20:30:02.14526 2017-01-13 12:30:02 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086587412660
2017-01-13_20:30:02.14527 2017-01-13 12:30:02 -0800 [debug]: plugin/buffer.rb:305:enqueue_chunk: enqueueing chunk instance=70086589215680 metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}>
2017-01-13_20:30:03.15465 2017-01-13 12:30:03 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086589414060
2017-01-13_20:30:03.15466 2017-01-13 12:30:03 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086588950900
2017-01-13_20:30:03.15467 2017-01-13 12:30:03 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086589215680
2017-01-13_20:30:03.15467 2017-01-13 12:30:03 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086587412660
2017-01-13_20:30:03.15467 2017-01-13 12:30:03 -0800 [debug]: plugin/buffer.rb:305:enqueue_chunk: enqueueing chunk instance=70086589215680 metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}>
2017-01-13_20:30:03.84340 2017-01-13 12:30:03 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086587269840
2017-01-13_20:30:04.16437 2017-01-13 12:30:04 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086588950900
2017-01-13_20:30:04.16440 2017-01-13 12:30:04 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086589414060
2017-01-13_20:30:04.16441 2017-01-13 12:30:04 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086587412660
2017-01-13_20:30:04.16441 2017-01-13 12:30:04 -0800 [debug]: plugin/buffer.rb:340:enqueue_all: enqueueing all chunks in buffer instance=70086589215680
2017-01-13_20:30:04.16441 2017-01-13 12:30:04 -0800 [debug]: plugin/buffer.rb:305:enqueue_chunk: enqueueing chunk instance=70086589215680 metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}>
2017-01-13_20:30:04.73929 2017-01-13 12:30:04 -0800 [debug]: plugin/out_kafka_buffered.rb:236:write: (records|bytes) ({"events"=>100}|{"events"=>143818})
2017-01-13_20:30:04.73945 2017-01-13 12:30:04 -0800 [debug]: plugin/buffer.rb:390:block in purge_chunk: purging a chunk instance=70086589215680 chunk_id="545ffaf0490f94f5ab7391b43eb50bea" metadata=#<struct Fluent::Plugin::Buffer::Metadata timekey=nil, tag=nil, variables={:key=>"events.production"}
```

It does look like there's some delay, but not enough delay that it shouldn't work at all?

**Comment by dtboctor** [comment]
So setting `kafka_agg_max_bytes = 128k` fixes the issue and I can use acks = -1, but you say this option will be gone soon?  Will there be a replacement for it or a way to continue to make this thing work based on my finding?

**Comment by dtboctor** [comment]
actually setting it to 8k also makes it work, looks like the default is 4k?

**Comment by jpaskhay** [comment]
FYI, we were running into a similar problem when using `required_acks` of anything other than 0. We discovered the bottleneck in the ruby-kafka plugin's instrumentation and skip the relevant code block in our forked repo to get around it. Refer to https://github.com/zendesk/ruby-kafka/issues/327

After skipping the code block referenced in that issue, we got about 3x throughput when using `required_acks` of 1 and even -1.

Hope this helps others!
